File: clients/omni-windows/src/Omni.App.Core/DaemonViewModel.cs
using System.Collections.ObjectModel;
using Omni.Ipc;

namespace Omni.App.Core;

/// <summary>
/// The app's main view model: a live, bindable view of the daemon. It performs the
/// version handshake, follows the daemon's push stream to keep state current,
/// reconnects on its own when the daemon goes away, and exposes commands that map
/// straight to IPC requests. It holds no business logic — the daemon owns state.
/// </summary>
public sealed class DaemonViewModel : ObservableObject
{
    private readonly IOmniDaemonClient _client;
    private readonly IUiDispatcher _ui;
    private readonly TimeSpan _reconnectDelay;

    public DaemonViewModel(IOmniDaemonClient client, IUiDispatcher ui, TimeSpan? reconnectDelay = null)
    {
        _client = client;
        _ui = ui;
        _reconnectDelay = reconnectDelay ?? TimeSpan.FromSeconds(2);

        // Keep the "has any …" flags (used to show/hide sections) in step with the
        // collections; these run on the UI thread, inside the same update.
        Pending.CollectionChanged += (_, _) => Raise(nameof(HasPending));
        Sessions.CollectionChanged += (_, _) => Raise(nameof(HasSessions));
        Peers.CollectionChanged += (_, _) => Raise(nameof(HasPeers));
        Placements.CollectionChanged += (_, _) => Raise(nameof(HasPlacements));
    }

    private ConnectionStatus _connection = ConnectionStatus.Connecting;
    public ConnectionStatus Connection
    {
        get => _connection;
        private set
        {
            if (SetField(ref _connection, value))
            {
                Raise(nameof(IsConnected));
                Raise(nameof(IsIncompatible));
            }
        }
    }

    /// <summary>True only while live, for enabling/disabling controls.</summary>
    public bool IsConnected => _connection == ConnectionStatus.Connected;

    /// <summary>True when the daemon is too new; the UI shows an "update" notice.</summary>
    public bool IsIncompatible => _connection == ConnectionStatus.Incompatible;

    private string _statusText = "Connecting…";
    public string StatusText { get => _statusText; private set => SetField(ref _statusText, value); }

    private string _fingerprint = "";
    public string Fingerprint { get => _fingerprint; private set => SetField(ref _fingerprint, value); }

    private int _port;
    public int Port { get => _port; private set => SetField(ref _port, value); }

    private bool _capturing;
    public bool Capturing { get => _capturing; private set => SetField(ref _capturing, value); }

    private bool _clipboardSharing;
    public bool ClipboardSharing { get => _clipboardSharing; private set => SetField(ref _clipboardSharing, value); }

    private string _daemonVersion = "";
    public string DaemonVersion { get => _daemonVersion; private set => SetField(ref _daemonVersion, value); }

    private string? _lastError;
    public string? LastError
    {
        get => _lastError;
        private set
        {
            if (SetField(ref _lastError, value))
            {
                Raise(nameof(HasError));
            }
        }
    }

    /// <summary>True when there is an error message to show.</summary>
    public bool HasError => !string.IsNullOrEmpty(_lastError);

    public ObservableCollection<SessionInfo> Sessions { get; } = [];
    public ObservableCollection<PendingInfo> Pending { get; } = [];
    public ObservableCollection<PeerInfo> Peers { get; } = [];
    public ObservableCollection<LayoutInfo> Placements { get; } = [];

    public bool HasPending => Pending.Count > 0;
    public bool HasSessions => Sessions.Count > 0;
    public bool HasPeers => Peers.Count > 0;
    public bool HasPlacements => Placements.Count > 0;

    /// <summary>
    /// Runs until <paramref name="cancellationToken"/> is cancelled: handshake,
    /// then follow the push stream, reconnecting after a delay whenever it drops.
    /// Stops permanently only if the daemon is too new (see
    /// <see cref="ConnectionStatus.Incompatible"/>).
    /// </summary>
    public async Task RunAsync(CancellationToken cancellationToken)
    {
        while (!cancellationToken.IsCancellationRequested)
        {
            try
            {
                var hello = await _client.HelloAsync(cancellationToken).ConfigureAwait(false);
                if (hello.ProtocolVersion > OmniProtocol.Version)
                {
                    SetState(ConnectionStatus.Incompatible,
                        $"The daemon speaks protocol v{hello.ProtocolVersion}; this app understands v{OmniProtocol.Version}. Please update Omnipresent.");
                    return;
                }
                _ui.Post(() => DaemonVersion = hello.DaemonVersion);

                await foreach (var snapshot in _client.SubscribeAsync(cancellationToken).ConfigureAwait(false))
                {
                    Apply(snapshot);
                    await RefreshListsAsync(cancellationToken).ConfigureAwait(false);
                    SetState(ConnectionStatus.Connected, "Connected");
                }
            }
            catch (OperationCanceledException)
            {
                break;
            }
            catch (OmniDaemonException ex)
            {
                SetState(ConnectionStatus.Disconnected, ex.Message);
            }

            if (cancellationToken.IsCancellationRequested)
            {
                break;
            }
            SetState(ConnectionStatus.Disconnected, "Waiting for the daemon…");
            try
            {
                await Task.Delay(_reconnectDelay, cancellationToken).ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                break;
            }
        }
    }

    public Task ConnectAsync(string host) => Guard(ct => _client.ConnectAsync(host, ct));
    public Task DisconnectAsync(string host) => Guard(ct => _client.DisconnectAsync(host, ct));
    public Task AcceptAsync(string selector) => Guard(ct => _client.AcceptAsync(selector, ct));
    public Task RejectAsync(string selector) => Guard(ct => _client.RejectAsync(selector, ct));
    public Task RemovePeerAsync(string selector) => Guard(ct => _client.RemovePeerAsync(selector, ct));
    public Task SetLayoutAsync(string host, string edge) => Guard(ct => _client.SetLayoutAsync(host, edge, ct));
    public Task SetClipboardAsync(bool enabled) => Guard(ct => _client.SetClipboardAsync(enabled, ct));

    private async Task Guard(Func<CancellationToken, Task> action)
    {
        _ui.Post(() => LastError = null);
        try
        {
            await action(CancellationToken.None).ConfigureAwait(false);
        }
        catch (OmniDaemonException ex)
        {
            _ui.Post(() => LastError = ex.Message);
        }
    }

    private void Apply(StatusInfo snapshot) => _ui.Post(() =>
    {
        Fingerprint = snapshot.Fingerprint;
        Port = snapshot.Port;
        Capturing = snapshot.Capturing;
        ClipboardSharing = snapshot.ClipboardSharing;
        Replace(Sessions, snapshot.Sessions);
        Replace(Pending, snapshot.Pending);
    });

    private async Task RefreshListsAsync(CancellationToken cancellationToken)
    {
        // Peers and placements are separate requests, not part of the snapshot;
        // refresh them whenever state changes so the lists stay live.
        try
        {
            var peers = await _client.PeersAsync(cancellationToken).ConfigureAwait(false);
            var placements = await _client.LayoutAsync(cancellationToken).ConfigureAwait(false);
            _ui.Post(() =>
            {
                Replace(Peers, peers);
                Replace(Placements, placements);
            });
        }
        catch (OmniDaemonException)
        {
            // Keep whatever we last had; the snapshot itself still applied.
        }
    }

    private void SetState(ConnectionStatus status, string text) => _ui.Post(() =>
    {
        Connection = status;
        StatusText = text;
        if (status == ConnectionStatus.Incompatible)
        {
            LastError = text;
        }
    });

    private static void Replace<T>(ObservableCollection<T> target, IReadOnlyList<T> items)
    {
        target.Clear();
        foreach (var item in items)
        {
            target.Add(item);
        }
    }
}

File: clients/omni-macos/omni-macos/Core/DaemonViewModel.swift
//
//  DaemonViewModel.swift
//  omni-macos
//
//  A live, bindable view of the daemon. It performs the version handshake,
//  follows the daemon's push stream to keep state current, reconnects on its
//  own when the daemon goes away, and exposes commands that map straight to IPC
//  requests. It holds no business logic — the daemon owns state. Mirrors the
//  Windows client's `DaemonViewModel`.
//

import Foundation
import Observation

@Observable
@MainActor
public final class DaemonViewModel {
    /// The client's view of the IPC connection.
    public enum ConnectionStatus {
        case connecting
        case connected
        case disconnected
        case incompatible
    }

    private let client: DaemonClient
    private let reconnectDelay: Duration
    private var runTask: Task<Void, Never>?

    public init(client: DaemonClient, reconnectDelay: Duration = .seconds(2)) {
        self.client = client
        self.reconnectDelay = reconnectDelay
    }

    /// Starts the connect/subscribe/reconnect loop once, for the whole app
    /// lifetime. Idempotent: safe to call from both the window and the menu-bar
    /// views, so live updates keep flowing even while the window is closed.
    public func start() {
        guard runTask == nil else { return }
        runTask = Task { await self.run() }
    }

    // MARK: - Observed state

    public private(set) var connection: ConnectionStatus = .connecting
    public private(set) var statusText = "Connecting…"
    public private(set) var fingerprint = ""
    public private(set) var port: UInt16 = 0
    public private(set) var capturing = false
    public private(set) var clipboardSharing = false
    public private(set) var daemonVersion = ""
    public private(set) var lastError: String?

    public private(set) var sessions: [SessionInfo] = []
    public private(set) var pending: [PendingInfo] = []
    public private(set) var peers: [PeerInfo] = []
    public private(set) var placements: [LayoutInfo] = []

    /// True only while live, for enabling/disabling controls.
    public var isConnected: Bool { connection == .connected }
    /// True when the daemon is too new; the UI shows an "update" notice.
    public var isIncompatible: Bool { connection == .incompatible }
    /// True when there is an error message to show.
    public var hasError: Bool { !(lastError ?? "").isEmpty }

    // MARK: - Lifecycle

    /// Runs until the surrounding task is cancelled: handshake, then follow the
    /// push stream, reconnecting after a delay whenever it drops. Stops
    /// permanently only if the daemon is too new (`incompatible`).
    public func run() async {
        while !Task.isCancelled {
            do {
                let hello = try await client.hello()
                if hello.protocolVersion > OmniProtocol.version {
                    connection = .incompatible
                    statusText = """
                        The daemon speaks protocol v\(hello.protocolVersion); this app \
                        understands v\(OmniProtocol.version). Please update Omnipresent.
                        """
                    lastError = statusText
                    return
                }
                daemonVersion = hello.daemonVersion

                for try await snapshot in client.subscribe() {
                    apply(snapshot)
                    await refreshLists()
                    connection = .connected
                    statusText = "Connected"
                }
            } catch is CancellationError {
                break
            } catch let error as OmniDaemonError {
                setDisconnected(error.message)
            } catch {
                setDisconnected(error.localizedDescription)
            }

            if Task.isCancelled { break }
            // Use .connecting (not .disconnected) while sleeping before the next
            // attempt — the daemon may just be starting up, so orange is correct.
            connection = .connecting
            statusText = "Waiting for the daemon…"
            do {
                try await Task.sleep(for: reconnectDelay)
            } catch {
                break  // cancelled while waiting
            }
        }
    }

    // MARK: - Commands

    public func connect(host: String) async { await guarded { try await self.client.connect(host: host) } }
    public func disconnect(host: String) async { await guarded { try await self.client.disconnect(host: host) } }
    public func accept(selector: String) async { await guarded { try await self.client.accept(selector: selector) } }
    public func reject(selector: String) async { await guarded { try await self.client.reject(selector: selector) } }
    public func removePeer(selector: String) async { await guarded { try await self.client.removePeer(selector: selector) } }
    public func setLayout(host: String, edge: String) async { await guarded { try await self.client.setLayout(host: host, edge: edge) } }
    public func setClipboard(enabled: Bool) async { await guarded { try await self.client.setClipboard(enabled: enabled) } }

    /// Launches `omni start` then immediately retries the connection so the UI
    /// turns green without waiting for the next scheduled reconnect interval.
    public func startDaemon() async {
        lastError = nil
        let candidates = [
            "/usr/local/bin/omni",
            "/opt/homebrew/bin/omni",
            "\(NSHomeDirectory())/.local/bin/omni",
            "\(NSHomeDirectory())/.cargo/bin/omni",
        ]
        guard let binaryPath = candidates.first(where: {
            FileManager.default.isExecutableFile(atPath: $0)
        }) else {
            lastError = "Could not find the omni binary. Make sure Omnipresent is installed."
            return
        }
        do {
            let process = Process()
            process.executableURL = URL(fileURLWithPath: binaryPath)
            process.arguments = ["start"]
            try process.run()
        } catch {
            lastError = "Failed to start the daemon: \(error.localizedDescription)"
            return
        }
        // Restart the run loop immediately so we don't wait for the retry sleep.
        reconnectNow()
    }

    /// Sends a stop command to the running daemon.
    public func stopDaemon() async {
        await guarded { try await self.client.stop() }
    }

    /// Cancels the current run loop and starts a fresh one immediately,
    /// bypassing any in-progress reconnect-delay sleep.
    public func reconnectNow() {
        runTask?.cancel()
        runTask = nil
        runTask = Task { await self.run() }
    }

    // MARK: - Internals

    private func guarded(_ action: () async throws -> Void) async {
        lastError = nil
        do {
            try await action()
        } catch let error as OmniDaemonError {
            lastError = error.message
        } catch {
            lastError = error.localizedDescription
        }
    }

    private func apply(_ snapshot: StatusInfo) {
        fingerprint = snapshot.fingerprint
        port = snapshot.port
        capturing = snapshot.capturing
        clipboardSharing = snapshot.clipboardSharing
        sessions = snapshot.sessions
        pending = snapshot.pending
    }

    private func refreshLists() async {
        // Peers and placements are separate requests, not part of the snapshot;
        // refresh them whenever state changes so the lists stay live. Keep the
        // last values if a refresh fails — the snapshot itself still applied.
        do {
            let freshPeers = try await client.peers()
            let freshPlacements = try await client.layout()
            peers = freshPeers
            placements = freshPlacements
        } catch {
            // keep whatever we last had
        }
    }

    private func setDisconnected(_ text: String) {
        connection = .disconnected
        statusText = text
    }
}
